File: Calc/CalcWidget.cpp
#include <QLineEdit>
#include <QPushButton>
#include <QVBoxLayout>
#include <QGridLayout>
#include <QButtonGroup>

#include "CalcWidget.h"
#include "BasicCalc.h"

const int CalcWidget::N_NUM_BUTTONS = 10;
const int CalcWidget::N_OP_BUTTONS = 4;

CalcWidget::CalcWidget(QWidget * parent) :
    QWidget(parent)
{
    const char * ops [] = {"+", "-", "*", "/"};
    QVBoxLayout * mainLayout = new QVBoxLayout(this);
    QGridLayout * calcLayout = new QGridLayout();
    
    calcLineEdit = new QLineEdit();
    calcLineEdit->setReadOnly(true);
    mainLayout->addWidget(calcLineEdit);
    
    QButtonGroup * numberGroup = new QButtonGroup(this);
    numberButtons = new QPushButton * [N_NUM_BUTTONS];
    for (int i = 0; i < N_NUM_BUTTONS; i++) {
        numberButtons[i] = new QPushButton(QString::number(i));
        numberGroup->addButton(numberButtons[i], i);
    }
    calcLayout->addWidget(numberButtons[7], 0, 0);
    calcLayout->addWidget(numberButtons[8], 0, 1);
    calcLayout->addWidget(numberButtons[9], 0, 2);
    calcLayout->addWidget(numberButtons[4], 1, 0);
    calcLayout->addWidget(numberButtons[5], 1, 1);
    calcLayout->addWidget(numberButtons[6], 1, 2);
    calcLayout->addWidget(numberButtons[1], 2, 0);
    calcLayout->addWidget(numberButtons[2], 2, 1);
    calcLayout->addWidget(numberButtons[3], 2, 2);
    calcLayout->addWidget(numberButtons[0], 3, 0);
    
    clearButton = new QPushButton("C");
    calcLayout->addWidget(clearButton, 3, 1);
    
    resultButton = new QPushButton("=");
    calcLayout->addWidget(resultButton, 3, 2);
    
    QButtonGroup * opGroup = new QButtonGroup(this);
    opButtons = new QPushButton * [N_OP_BUTTONS];
    for (int i = 0; i < N_OP_BUTTONS; i++) {
        opButtons[i] = new QPushButton(ops[i]);
        opGroup->addButton(opButtons[i], i);
        calcLayout->addWidget(opButtons[i], N_OP_BUTTONS - 1 - i, 4);
    }
    mainLayout->addLayout(calcLayout);
    
    calc = new BasicCalc(this);
    
    connect(numberGroup, SIGNAL(buttonClicked(int)), calc, SLOT(putDigit(int)));
    connect(clearButton, SIGNAL(clicked()), calc, SLOT(clear()));
    connect(opGroup, SIGNAL(buttonClicked(int)), calc, SLOT(saveOperation(int)));
    connect(resultButton, SIGNAL(clicked()), calc, SLOT(computeResult()));
    
    connect(calc, SIGNAL(operandChanged(const QString &)),
            calcLineEdit, SLOT(setText(const QString &)));
}

CalcWidget::~CalcWidget()
{
    if (numberButtons != 0) {
        delete [] numberButtons;
    }
    if (opButtons != 0) {
        delete [] opButtons;
    }
}

File: Calc/python/CalcWidget.py
# encoding: utf-8

from PyQt4 import QtGui
from BasicCalc import BasicCalc

class CalcWidget(QtGui.QWidget):

    ops = ['+', '-', '*', '/']
    N_NUM_BUTTONS = 10
    N_OP_BUTTONS = 4

    def __init__(self, parent = None):
        super(QtGui.QWidget, self).__init__(parent)
        mainLayout = QtGui.QVBoxLayout(self)
        calcLayout = QtGui.QGridLayout()
        
        self.calcLineEdit = QtGui.QLineEdit()
        self.calcLineEdit.setReadOnly(True)
        mainLayout.addWidget(self.calcLineEdit)
        
        numberGroup = QtGui.QButtonGroup(self)
        self.numberButtons = []
        for i in range(0, self.N_NUM_BUTTONS):
            self.numberButtons.append(QtGui.QPushButton(str(i)))
            numberGroup.addButton(self.numberButtons[i], i)
        calcLayout.addWidget(self.numberButtons[7], 0, 0)
        calcLayout.addWidget(self.numberButtons[8], 0, 1)
        calcLayout.addWidget(self.numberButtons[9], 0, 2)
        calcLayout.addWidget(self.numberButtons[4], 1, 0)
        calcLayout.addWidget(self.numberButtons[5], 1, 1)
        calcLayout.addWidget(self.numberButtons[6], 1, 2)
        calcLayout.addWidget(self.numberButtons[1], 2, 0)
        calcLayout.addWidget(self.numberButtons[2], 2, 1)
        calcLayout.addWidget(self.numberButtons[3], 2, 2)
        calcLayout.addWidget(self.numberButtons[0], 3, 0)
        
        self.clearButton = QtGui.QPushButton("C")
        calcLayout.addWidget(self.clearButton, 3, 1)
        
        self.resultButton = QtGui.QPushButton("=")
        calcLayout.addWidget(self.resultButton, 3, 2)
        
        opGroup = QtGui.QButtonGroup(self)
        self.opButtons = []
        for i in range(0, self.N_OP_BUTTONS):
            self.opButtons.append(QtGui.QPushButton(self.ops[i]))
            opGroup.addButton(self.opButtons[i], i)
            calcLayout.addWidget(self.opButtons[i], self.N_OP_BUTTONS - 1 - i, 4)
            
        mainLayout.addLayout(calcLayout)
        
        self.calc = BasicCalc(self)
        numberGroup.buttonClicked[int].connect(self.calc.putDigit)
        self.clearButton.clicked.connect(self.calc.clear)
        opGroup.buttonClicked[int].connect(self.calc.saveOperation)
        self.resultButton.clicked.connect(self.calc.computeResult)
        self.calc.operandChanged.connect(self.calcLineEdit.setText)

